### PR TITLE
Make configuration (and therefore builds) leaner

### LINKED ISSUE #16377: A bit too much rebuilding when reconfiguration didn't reconfigure much
Now that we've figured out a few items where dependencies didn't quite catch (or caught too much as the case may be), fixed by #16361 and #16365, we now have a situation with unnecessary rebuilding whenever Configure is re-run.

This is due to the files `configdata.pm` being rewritten and thereby getting its time stamp updated, even though the contents may not have changed.  Many other files (especially generated ones) depend on this file, and are thereby regenerated as well.

Among the files that are generated (and depend on `configdata.pm`) is `include/openssl/configuration.h`, which *more or less all of our source code* depends on, directly or via other header files.

For anyone that wants to quickly build the latest development on the master branch, it comes as a bit of a surprise.
For any automation (such as buildbot) where incremental builds are desired to get quick turnaround time for developers, this becomes hugely wasted time.

This isn't very hard to solve.  As a matter of fact, we mostly have what's needed in place to correct this, most of what's needed is a simple file compare and to back off a newly produced file if its contents don't differ from the old file.

### PR DESCRIPTION
**Pull request by levitte**
To make incremental builds more possible, we need to have re-configuration impact building as little as possible.  Since rebuilding is sensitive to time stamps, that becomes the important thing to keep track of.

Two files are affected, `configdata.pm` and `include/openssl/configuration.h`.  With this PR, we avoid having their time stamps updated unless their contents have changed, and thusly avoid rebuilding stuff that depend on them unless necessary.

Please read the actual commit messages for the full story.

Fixes #16377

### CODE REVIEW

**Comment by paulidale** [comment]
How safe is this for 3.0?  It seems more like a feature to me but it could be argued both ways easily enough.

**Comment by levitte** [comment]
> How safe is this for 3.0? It seems more like a feature to me but it could be argued both ways easily enough.

As far as I can determine, it's safe enough.  This is really doing the exact same thing as before, but moved to a different spot

**Comment by mattcaswell** [comment]
IMO this should go into 3.0

**Comment by richsalz** [comment]
If you don't take this PR, then you should revert the ones that made it necessary.  Needlessly rebuilding is really bad.

This PR is serious magic tho. I would have expected something like GNU's `mv-if-changed` somewhere.

**Comment by levitte** [comment]
> This PR is serious magic tho. I would have expected something like GNU's `mv-if-changed` somewhere.

That won't help, because at least *some* make implementations will consider that if a make target is called, it's updated.  So from a building perspective, this doesn't help anything:

``` make
foo.txt:
	echo something > foo.txt.new
	mv-if-changed foo.txt.new foo.txt
	rm -f foo.txt.new

whatever.txt: foo.txt
	date > whatever.txt
```

Doing `make whatever.txt` will do `date > whatever.txt` every time, regardless if foo.txt was actually changed or not.

**Comment by levitte** [comment]
Regarding GNU, I think there are very good reasons why they have `AC_CONFIG_HEADERS`.  They got to learn this lesson a long time ago.

**Comment by levitte** [comment]
> If you don't take this PR, then you should revert the ones that made it necessary. Needlessly rebuilding is really bad.

"the ones" that I think you refer to fixed a *separate* problem.  That separate problem just happened to cover up the issue that this PR fixes, probably mostly because it forced everyone to do a `make clean; perl configdata.pm -r` a little now and then anyway.

**Comment by levitte** [comment]
Would everyone be more comfortable if we had a common `mv-if-changed` like function in OpenSSL::Util (`util/perl/OpenSSL/Util.pm`)?  That should be too hard to do...

**Comment by mattcaswell** [comment]
It's unclear what is happening with this PR. It seems to have stalled. Is it still relevant/wanted?

**Comment by levitte** [comment]
Thanks for the reminder.  I'll get back to it.

I would like this to actually happen, as it promotes incremental builds better

File: configdata.pm.in [hunk 1]
```diff
@@ -131,10 +132,10 @@ use lib '{- $config{builddir} -}';
 use platform;
 _____
 
-        print 'Creating ',$buildfile,"\n";
+        my $tmpl;
```
Comment (tmshort) [inline configdata.pm.in:135]: [Nit] Why is this moved up here?

Comment (levitte) [reply configdata.pm.in:135]: This is a stylistic thing, to make it more apparent where this variable is defined, since it's assigned a new value down in (new) line 160.

File: configdata.pm.in [hunk 2]
```diff
@@ -148,6 +149,57 @@ _____
...
+        my $update_configuration_h = 0;
+        if (-f $configuration_h) {
+            my $configuration_h_mtime = (stat($configuration_h))[9];
+            my $configuration_h_in_mtime = (stat($configuration_h_in))[9];
+
+            # If configuration.h.in was updated after the last configuration.h,
+            # or if configuration.h.new differs configuration.h, we update
+            # configuration.h
+            if ($configuration_h_mtime < $configuration_h_in_mtime
+                || compare_text("$configuration_h.new", $configuration_h) != 0) {
```
Comment (tmshort) [inline configdata.pm.in:187]: [Nit], This could be misread as "$configuration.h.new"; I'm a fan of using braces around variables in strings:
```
"${configuration_h}.new"

**Review by tmshort — APPROVED** [review]
Just some minor nits that don't necessarily need to be fixed.

It's probably out-of-scope to put in an error or warning about target change (as described in #17979).

File: configdata.pm.in (hunk 2 above)
Comment (levitte) [reply configdata.pm.in:187]: Fixed

**Comment by levitte** [comment]
> It's probably out-of-scope to put in an error or warning about target change (as described in #17979).

Yeah

**Comment by levitte** [comment]
@openssl/otc, review please

**Review by t8m — APPROVED** [review]

**Comment by levitte** [comment]
Merged

master:
764cf5b26306a8712e8b3d41599c44dc5ed07a25 Configuration: only produce a new configdata.pm if it has changed contents
2522889620446f1e56338367d1b6b028ea952bb4 Configuration: produce include/openssl/configuration.h when configuring
4d02d500aac80c136e3d6582b908e0fab77bbf42 Building: For the FIPS module checksum, keep track of configuration,h

3.0:
7d78017bd88758cc4cc1337da282f5439c17719a Configuration: only produce a new configdata.pm if it has changed contents
5cbd580be9012edb39a60dc303f9ea55c7e4d7bd Configuration: produce include/openssl/configuration.h when configuring
ff3e71b33c3f0f2da189b102afe13e074c85646a Building: For the FIPS module checksum, keep track of configuration,h

**Comment by levitte** [comment]
This turns out to partly be a mistake.

The commit that has `configdata.pm` produce `include/openssl/configuration.h` alongside `Makefile` is fine.

The commit that ensure the FIPS checksum depends on `include/openssl/configuration.h` is fine.

*The commit that has `Configure` and `configdata.pm` do the comparison magic is **NOT** fine*, as it competes with the dependencies for `Makefile` and `configdata.pm` that are set in `Makefile`.
A perfect example is now in 3.0, where `Configurations/10-main.conf` has changed, and this is detected by `make` (because of the dependencies layed out in `Makefile`), which then tries to reconfigure, only to end up with an untouched `configdata.pm`, and that has the caller stuck with a `make` that tries to reconfigure to no avail.
Lesson learned: don't try to outsmart `make`.

I'm going to submit a reversal PR as soon as I can.

**Comment by levitte** [comment]
As evidence for the problem I just mentioned, here's the listing of relevant files and relevant dependencies in a build I have:

``` console
$ ls -l configdata.pm* Makefile Configurations/10-main.conf 
$ ls -ltr configdata.pm* Makefile Configurations/10-main.conf 
-rw-r--r-- 1 buildbot buildbot   16738 Jun  1 08:32 configdata.pm.in
-rwxr-xr-x 1 buildbot buildbot 1124600 Jul 14 14:11 configdata.pm
-rw-r--r-- 1 buildbot buildbot 5265589 Jul 14 14:11 Makefile
-rw-r--r-- 1 buildbot buildbot   82392 Jul 18 09:11 Configurations/10-main.conf
$ grep -A1 ^Makefile: Makefile
Makefile: configdata.pm \
          Configurations/common0.tmpl Configurations/unix-Makefile.tmpl
$ grep -A54 ^configdata.pm: Makefile
configdata.pm: $(SRCDIR)/Configure $(SRCDIR)/config \
               ./build.info crypto/build.info ssl/build.info apps/build.info \
               util/build.info tools/build.info fuzz/build.info \
               providers/build.info doc/build.info test/build.info \
               engines/build.info crypto/objects/build.info \
               crypto/buffer/build.info crypto/bio/build.info \
               crypto/stack/build.info crypto/lhash/build.info \
               crypto/rand/build.info crypto/evp/build.info \
               crypto/asn1/build.info crypto/pem/build.info \
               crypto/x509/build.info crypto/conf/build.info \
               crypto/txt_db/build.info crypto/pkcs7/build.info \
               crypto/pkcs12/build.info crypto/ui/build.info \
               crypto/kdf/build.info crypto/store/build.info \
               crypto/property/build.info crypto/md4/build.info \
               crypto/md5/build.info crypto/sha/build.info \
               crypto/mdc2/build.info crypto/hmac/build.info \
               crypto/ripemd/build.info crypto/whrlpool/build.info \
               crypto/poly1305/build.info crypto/siphash/build.info \
               crypto/sm3/build.info crypto/des/build.info \
               crypto/aes/build.info crypto/rc2/build.info \
               crypto/rc4/build.info crypto/idea/build.info \
               crypto/aria/build.info crypto/bf/build.info \
               crypto/cast/build.info crypto/camellia/build.info \
               crypto/seed/build.info crypto/sm4/build.info \
               crypto/chacha/build.info crypto/modes/build.info \
               crypto/bn/build.info crypto/ec/build.info \
               crypto/rsa/build.info crypto/dsa/build.info \
               crypto/dh/build.info crypto/sm2/build.info \
               crypto/dso/build.info crypto/engine/build.info \
               crypto/err/build.info crypto/comp/build.info \
               crypto/http/build.info crypto/ocsp/build.info \
               crypto/cms/build.info crypto/ts/build.info \
               crypto/srp/build.info crypto/cmac/build.info \
               crypto/ct/build.info crypto/async/build.info \
               crypto/ess/build.info crypto/crmf/build.info \
               crypto/cmp/build.info crypto/encode_decode/build.info \
               crypto/ffc/build.info apps/lib/build.info \
               providers/common/build.info \
               providers/implementations/build.info doc/man1/build.info \
               providers/common/der/build.info \
               providers/implementations/digests/build.info \
               providers/implementations/ciphers/build.info \
               providers/implementations/rands/build.info \
               providers/implementations/macs/build.info \
               providers/implementations/kdfs/build.info \
               providers/implementations/exchange/build.info \
               providers/implementations/keymgmt/build.info \
               providers/implementations/signature/build.info \
               providers/implementations/asymciphers/build.info \
               providers/implementations/encode_decode/build.info \
               providers/implementations/storemgmt/build.info \
               providers/implementations/kem/build.info \
               providers/implementations/rands/seeding/build.info \
               Configurations/00-base-templates.conf \
               Configurations/10-main.conf
```

Especially, pay attention to the date of `Configurations/10-main.conf`.  This is due to the update from #18807, a new config target that doesn't change my configuration at all...  but with the dependencies above, it still triggers an endless reconfiguration.

**Comment by levitte** [comment]
See #18832
